apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
-                                    {
-                                        _ackQueue.Enqueue(msg.AckId);
-                                        triggerQEvent = true;
-                                    }
-                                    else
-                                    {
-                                        _unAckedMsgCount -= 1;
-                                    }
-                                }
-                                if (triggerQEvent)
-                                {
-                                    // Unblock Push(), because now there is at least one ACK to push.
-                                    _qEvent.Set();
-                                }
-                            }));
-                        }
+                        HandleRpcFailure(e);
+                        return;
                     }
+                    var receivedMessages = current.ReceivedMessages;
+                    var msgs = receivedMessages.ToList();
+                    receivedMessages.Clear();
+                    // Get all ack-ids, used to extend leases as required.
+                    var msgIds = new HashSet<string>(msgs.Select(x => x.AckId));
+                    // Send an initial "lease-extension"; which starts the server timer.
+                    HandleExtendLease(msgIds);
+                    // Asynchonously start message processing. Handles flow, and calls the user-supplied message handler.
+                    // Uses Task.Run(), so not to clog up this "master" thread with per-message processing.
+                    Task messagesTask = _taskHelper.Run(() => ProcessPullMessagesAsync(msgs, msgIds));
+                    // Once all received messages have been queued for processing, read the stream for more messages.
+                    Add(messagesTask, Next(true, () => HandlePullMoveNext(null)));
+                }
+                else
+                {
+                    StopStreamingPull();
+                    // Always a short pause on server disconnect.
+                    _pullBackoff = TimeSpan.FromSeconds(0.5);
+                    StartStreamingPull();
                 }
             }
 
-            private void QueueLeaseExtensions(IEnumerable<string> ackIds)
+            private async Task ProcessPullMessagesAsync(List<ReceivedMessage> msgs, HashSet<string> msgIds)
             {
-                lock (_qLock)
+                // Running async. Common data needs locking
+                for (int msgIndex = 0; msgIndex < msgs.Count; msgIndex++)
                 {
-                    _extendQueue.EnqueueAll(ackIds);
+                    _softStopCts.Token.ThrowIfCancellationRequested();
+                    var msg = msgs[msgIndex];
+                    msgs[msgIndex] = null;
+                    await _taskHelper.ConfigureAwait(_flow.Process(msg.CalculateSize(), async () =>
+                    {
+                            // Running async. Common data needs locking
```

Comment:
Indentation within this block has gone a bit screwy.

Reply:
Oops, done.